File: src/main/java/str/StrStr.java
package str;

import java.util.Arrays;

/**
 * @author comyxy
 */
public class StrStr {
    /**
     * Sunday
     *
     * @param haystack string
     * @param pattern  pattern
     * @return index of the pattern in string
     */
    public int sunday(String haystack, String pattern) {
        int n = pattern.length();
        if (n == 0) {
            return 0;
        }
        int m = haystack.length();
        // 建立偏移表
        int[] offset = new int[26];
        Arrays.fill(offset, n + 1);
        for (int i = 0; i < n; i++) {
            offset[pattern.charAt(i) - 'a'] = n - i;
        }

        // sunday
        int idx = 0;
        while (idx + n <= m) {
            String s = haystack.substring(idx, idx + n);
            if (pattern.equals(s)) {
                return idx;
            }
            if (idx + n >= m) {
                return -1;
            }
            char nc = haystack.charAt(idx + n);
            idx += offset[nc - 'a'];
        }
        return -1;
    }
}

File: src/main/java/str/StrStr.kt
package str

/**
 * @since 2021/8/4 22:52
 */
class Sunday {
    fun sunday(haystack: String, pattern: String): Int {
        val n = pattern.length
        if (n == 0) {
            return 0;
        }
        val m = haystack.length
        // haystack中字符不在pattern中 右移m+1
        val offset = IntArray(26) { _ -> n + 1 }
        // haystack中字符在pattern中 右移(n-i)
        for (i in 0 until n) {
            offset[pattern[i] - 'a'] = n - i
        }
        //
        var idx = 0
        while (idx <= m - n) {
            val sub = haystack.substring(idx, idx + n)
            if (sub == pattern) {
                return idx
            }
            if (idx + n >= m) {
                return -1
            }
            val c = haystack[idx + n]
            idx += offset[c - 'a']
        }
        return -1
    }
}